### ISSUE TITLE
Terminal.App notcurses-info output is a mess

### ISSUE THREAD

**Issue by dankamongmen** [dankamongmen/notcurses#1945]
see #1944. Terminal.App appears to be in pretty bad shape:

![terminalapp](https://user-images.githubusercontent.com/143473/125997854-c2da165a-5cee-48cb-bae0-5c64ad1a901e.png)

this bug was initially reported by @michaelsbradleyjr, who is a fucking champ .

**Comment by dankamongmen** [comment]
i've dialed back the log level associated with that "prohibited or invalid" output. @michaelsbradleyjr , does it still look this terrible? we want to get on top of the stderr problem, but i'd sleep easier knowing there isn't some distinct horrible issue with `Terminal.app`.

**Comment by dankamongmen** [comment]
erp we might have wanted to fix this before submitting a formula to brew =[

**Comment by michaelsbradleyjr** [comment]
It looks more interesting now and triggers the terminal bell from two to five times:

![image](https://user-images.githubusercontent.com/194260/127536854-b80b3cd7-835c-4dd7-b769-230fefe5fbd5.png)

Maybe add a note in the readme for macOS users: "Support for macOS is still a work-in-progress. For best results use Kitty or iTerm2, Terminal.app is known to not display expected output."

Let me know how I can help. As you know, I'm all thumbs with C and lacking some knowledge re: how terminal emulators work, but if you have an idea re: something I can poke at, I'll do my best.

**Comment by dankamongmen** [comment]
mother of god.

> For best results use Kitty or iTerm2, Terminal.app is known to not display expected output

this is one step above abandoning Software Witchery and becoming a french mime. better to take on the mantle of Mike Tyson. Terminal.App, my style is impetuous. My defense is impregnable, and I'm just ferocious. I want your heart, I want to eat your children!

**Comment by dankamongmen** [comment]
> this is one step above abandoning Software Witchery and becoming a french mime. better to take on the mantle of Mike Tyson. Terminal.App, my style is impetuous. My defense is impregnable, and I'm just ferocious. I want your heart, I want to eat your children!

praise be to allah! ﷽

**Comment by michaelsbradleyjr** [comment]
🥊

**Comment by dankamongmen** [comment]
=] if you've never read tyson's autobiography, it is in complete seriousness a stunning document.

alrightie i'm gonna get out my cheap macintosh and see what can be done, now that notcurses is in brew. as for how you might help, i mean absolutely you can. you needn't know c, but you do need know how terminals work. but here is what you can do, if you want to:

* run it in iTerm2 or somewhere where you're confident it's showing correctish output
* run it in Terminal.App and determine the failure points
* run e.g. `notcurses-info | xxd -e` (or just `notcurses-info > erp` and open `erp` in an editor) in both and see where the differences are in actual output. do they seem right?
* take a look at `infocmp -a` in Terminal.App. what all is missing that i'm perhaps trying to use?
* aside: it looks like i'm trying to display the logo using cell blitting if pixel blitting is not found, which i definitely don't intend to be doingm, whoops?
* if there's something correct i ought be emitting, what is it? `echo -e '\e[balhblah` etc to test

**Comment by dankamongmen** [comment]
also, it looks like possibly Terminal.App is reporting cursor position inverted (like `kmscon` does). if you run `notcurses-info` at various points within the screen vertically, does it always return to the top? if so, that's probably happening.

**Comment by michaelsbradleyjr** [comment]
~~I'm not sure~~ I wasn't sure if it's returning to the top.

When running `notcurses-info` for the first time, there's the output and the terminal bells (varying in number), and then the input line in the shell is filled with garbage.

If I backspace out the garbage, and run `notcurses-info` again, I see this:

![image](https://user-images.githubusercontent.com/194260/127723705-d8c7e8f5-62ff-4bd8-85bd-1c3a3d1e7863.png)

Maybe it did return to the top, a little hard to say...

Ah, on a subsequent run o `notcurses-info`, there were no new lines added, so I guess it did return to the top.

![image](https://user-images.githubusercontent.com/194260/127723813-49ccbcbe-ec3a-466b-8e3c-687c0dbee2b5.png)

**Comment by dankamongmen** [comment]
yep, that pretty much confirms my suspicions. it's printing the banner where you start, and then moving to the top of the screen to print the rest. (everything from "Notcurses 2.3.12..." through "Specify a (correct)..." comes from `init_banner()`). that's because after printing the init banner, i do a cursor location check (the reasons why are complicated), and it's almost certainly reporting 1 as the row, when it means column. i then go to row 1 to do more printing.

what happens if you run `read -p $'\e[6n' -dR`? about halfway down my terminal, i get:

```
[schwarzgerat](0) $ read -p $'\e[6n' -dR
^[[26;1R[schwarzgerat](0) $ 
```

**Comment by dankamongmen** [comment]
so i also see a few instances here where e.g. our background rgb query is being leaked as regular input, rather than consumed. what to do about that, hrmmm....

**Comment by dankamongmen** [comment]
ok, the big image on the right ought no longer be displayed,

**Comment by michaelsbradleyjr** [comment]
> ok, the big image on the right ought no longer be displayed,

It is no longer displayed.

![image](https://user-images.githubusercontent.com/194260/127742501-bdb5423e-7abf-467b-9827-c5f0aecce8fc.png)

> what happens if you run `read -p $'\e[6n' -dR`

Re: the terminal in the screenshot above, if I backspace out the garbage in the input line and run it, on the next line I get:
```
^[[25;1Rbash-5.1$
```

**Comment by dankamongmen** [comment]
hrmm, well nevermind, the cursor is being reported in the correct order.

so, this mac proves to be almost entirely worthless. homebrew cannot be installed; safari won't connect to most web sites. i'm going to give it away. how do you feel about some remote desktopping?

**Comment by dankamongmen** [comment]
alright, here's what we're gonna do to get rid of the cruft you must backspace over: when `TERM_PROCESS` or whatver is defined to be `Apple_Terminal`, we'll circumvent everything else except the XTVERSION check (because who knows whether iTerm clear/resets TERM_PROCESS when started from a Terminal.App).

**Comment by dankamongmen** [comment]
theoretically it's possible that it can eat each individual query, but is returning them in bursts, and we're only reading the first one or something. to know conclusively, we'd need write each one individually from the command line. bah

**Comment by dankamongmen** [comment]
> theoretically it's possible that it can eat each individual query, but is returning them in bursts, and we're only reading the first one or something. to know conclusively, we'd need write each one individually from the command line. bah

well i can do that on my piece of shit here, assuming Terminal.App hasn't changed much since mountain lion, which i'm kinda doubting it really has

**Comment by michaelsbradleyjr** [comment]
I think I can configure the Mac mini so you could connect to it as a remote desktop. I need to investigate if/how one connects *from* a non-mac computer.

**Comment by dankamongmen** [comment]
it looks like it has a vnc server that one can tunnel under ssh

**Comment by michaelsbradleyjr** [comment]
It does have a built-in remote desktop ability, not sure if it's vnc/-compatible, maybe it is.

I'll try using my ubuntu laptop to connect, and if that works you should be able to do it over ssh with port forwarding, as you suggested... time to hook up the monitor and keyboard/mouse again, will give an update soon.

**Comment by dankamongmen** [comment]
if mountain lion's client will work, there's also that option.

i just did some experimentation and i think i might be onto something with my second theory -- it's replying in bursts, and we're not sewing them together for whatever reason. most of the cruft we see in your output is *replies*, not *queries*.

**Comment by michaelsbradleyjr** [comment]
You should be good to go:
```
ssh -L 5900:localhost:5900 ...
```

**Comment by dankamongmen** [comment]
i'm now on the macmini via vnc, thanks much @michaelsbradleyjr !

so one thing i note is that `notcurses-demo` works pretty well, more or less, even in Terminal.app. our big problems are just the moving to the top of the screen in `notcurses-info`, and all this crap coming back on our stdin after we run anything.

**Comment by dankamongmen** [comment]
so if i redirect `notcurses-info` stdout to a file, we see quite plainly that there's a definite `\e[1;1H` right after the initial banner. so we're definitely moving to the top of the screen by choice.

**Comment by dankamongmen** [comment]
i'm guessing that the inability to get the cursor location has a lot to do with the absolute minimal handling of replies that we're doing, which also leads to all the cruft on stdin.

![2021-08-02-031856_1524x1417_scrot](https://user-images.githubusercontent.com/143473/127819938-37973968-f2e5-4f3e-ad6c-f9f951ea87be.png)

**Comment by dankamongmen** [comment]
interestingly, the `direct` PoC reads cursor positions just fine.

**Comment by dankamongmen** [comment]
huh. not quite sure how much time had passed between this last checkout and now, but `notcurses-info` seems to pretty much completely work as of current tip-of-master. even shows some emoji.

**Comment by dankamongmen** [comment]
![2021-08-02-032821_1825x1301_scrot](https://user-images.githubusercontent.com/143473/127821053-e27eb253-e7bd-47fc-b9f5-48582850b9ed.png)

w00t?

**Comment by dankamongmen** [comment]
scrolling case also looks fine

![2021-08-02-032908_1825x1301_scrot](https://user-images.githubusercontent.com/143473/127821178-794eb1f8-00cf-4fd8-b79d-3aa20032b54b.png)

**Comment by dankamongmen** [comment]
i think we're done here?

i do notice that `notcurses-demo` doesn't want to run in a default iTerm2 window, because the orca is too big. i'll open a new bug on that. but Terminal.app seems fine. yay! @michaelsbradleyjr , please confirm, if possible.

**Comment by michaelsbradleyjr** [comment]
I realize this issue is long closed, but I did just try 2.3.17 in Terminal.app, iTerm2, and Kitty on macOS 10.15.

`notcurses-demo` and `notcurses-info` run fine in all three, with expected visual differences. There seems to be some problems with `ncneofetch`, but I'll try with 2.3.18 if/when I have time to build it locally later today.
